rtos/rtx5/mbed_rtx_idle.cpp

```diff
@@ -0,0 +1,198 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "rtos/rtos_idle.h"
+#include "platform/mbed_sleep.h"
+#include "TimerEvent.h"
+#include "us_ticker_api.h"
+#include "core_cm.h"
+#include "mbed_critical.h"
+#include "mbed_assert.h"
+#include <new>
+#include "rtx_os.h"
+extern "C" {
+#include "rtx_lib.h"
+}
+
+using namespace mbed;
+
+#if (defined(NO_SYSTICK))
+extern "C" IRQn_Type mbed_get_m0_tick_irqn(void);
+#endif
+
+class RtosTimer : private TimerEvent {
+public:
+    RtosTimer(): TimerEvent(get_us_ticker_data()), start_time(0), tick(0) {
+        start_time = ticker_read_us(_ticker_data);
+#if (defined(NO_SYSTICK))
+        NVIC_SetVector(mbed_get_m0_tick_irqn(), (uint32_t)SysTick_Handler);
+        NVIC_SetPriority(mbed_get_m0_tick_irqn(), 0xFF); /* RTOS requires lowest priority */
+        NVIC_EnableIRQ(mbed_get_m0_tick_irqn());
+#else
+        // Ensure SysTick has the correct priority as it is still used
+        // to trigger software interrupts on each tick. The period does
+        // not matter since it will never start counting.
+        SysTick_Setup(16);
+#endif
+    };
+
+    /**
+     * Schedule an os tick to fire
+     *
+     * @param delta Tick to fire at relative to current tick
+     */
+    void schedule_tick(uint32_t delta=1) {
+        insert_absolute(start_time + (tick + delta) * 1000);
+    }
+
+
+    /**
+     * Prevent any scheduled ticks from triggering
+     */
+    void cancel_tick() {
+        remove();
+    }
+
+    /**
+     * Get the current tick count
+     *
+     * @return The number of ticks since boot. This should match RTX's tick count
+     */
+    uint32_t get_tick() {
+        return tick & 0xFFFFFFFF;
+    }
+
+    /**
+     * Update the internal tick count
+     *
+     * @return The number of ticks incremented
+     */
+    uint32_t update_tick() {
+        uint64_t new_tick = ticker_read_us(_ticker_data) / 1000;
+        if (new_tick > tick) {
+            // Don't update to the current tick.
+            // The RTOS must do this to properly
+            // schedule tasks again.
+            new_tick--;
```

Comment:
It's not clear to me why we do `-1` in case `newtick > tick` ? If ticks have increased since the last update tick, we substract one from the tick count?

Reply:
The function svcRtxKernelResume doesn't trigger a task switch if a thread is ready to run. To work around this and jump start scheduling after a tick has passed, one less than the number of ticks that have occurred are passed to svcRtxKernelResume and SysTick is set to pending. The SysTick interrupt then increments the os's tick count to the right value and performs task switching as normal.

After further analysis, it appears that PendSV may be more appropriate in this situation. It will trigger a thread switch without recomputing the tick count. I'll update the PR to use this instead.